File: q1603/fib2.rb
Zero=[].size
One=[[]].size
Two=One<<One
Three=Two|One
Four=Two<<One
Five=Four|One
Six=Four|Two
Seven=Four|Two|One
Eight=Four<<One
Nine=Eight|One
Ten=Eight|Two
ADDMSK=(One<<Six).times.reduce(Zero){|s,i|(s<<One)|One}

def add(a,b)
return a+b
if true
	while b!=Zero
		c=(a&b)<<One
		a^=b
		b=c&ADDMSK
	end
else
	bchk=b
	msk=One
	c=Zero
	while (bchk|c)!=Zero
		cx=((a&b&msk)^(b&c)^(c&a))
		a^=(b&msk)^c
		c=cx
		c<<=One
		msk<<=One
		bchk>>=One
		c&=ADDMSK
	end
end
	a
end
def sub(a,b)
	add(~add(~Zero,b),a)
end

Mo=One|Two|Four|One<<Nine|One<<add(One,Ten)|One<<add(Four,Ten)|One<<add(Five,Ten)|One<<add(Seven,Ten)|One<<add(Nine,Ten)|One<<add(Ten,Ten)|One<<add(add(Three,Ten),Ten)|One<<add(add(Four,Ten),Ten)|One<<add(add(Five,Ten),Ten)|One<<add(add(Seven,Ten),Ten)|One<<add(add(Eight,Ten),Ten)|One<<add(add(Nine,Ten),Ten)

def mul(a,b)
	ret=Zero
	while b>Zero
		ret=add(ret,a) if (b&One)>Zero
		a=add(a,a)
		b>>=One
	end
	ret
end

def div(a,b)
	ret=Zero
	while a>=b
		i=Zero
		while a>=b<<i
			i=add(i,One)
		end
		i=sub(i,One)
		a=sub(a,b<<i)
		ret|=One<<i
	end
	ret
end

def mod(a,b)
	sub(a,mul(div(a,b),b))
end

def M(x,y)
	a=Zero;b=One;c=Two;d=Three;e=Four
	[
		mod(add(mul(x[a],y[a]),mul(x[b],y[c])),Mo),
		mod(add(mul(x[a],y[b]),mul(x[b],y[d])),Mo),
		mod(add(mul(x[c],y[a]),mul(x[d],y[c])),Mo),
		mod(add(mul(x[c],y[b]),mul(x[d],y[d])),Mo),
	]
end

gets.to_i.times{
	n=sub(gets.to_i,One)
	e=[One,Zero,Zero,One]
	x=[One,One,One,Zero]
	while n>Zero
		e=M(e,x) if (n&One)>Zero
		x=M(x,x)
		n>>=One
	end
	p mod(add(e[Zero],e[One]<<One),Mo)
}

File: q1603/fib2.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
using namespace std;

typedef vector<long long>V;
const long long zero=false;
const long long one=true;
const long long two=one<<one;
const long long three=two|one;
const long long four=two<<one;
const long long five=four|one;
const long long six=four|two;
const long long seven=four|two|one;
const long long eight=four<<one;
const long long nine=eight|one;
const long long ten=eight|two;

constexpr long long add(long long a,long long b){
	long long bchk=b;
	long long msk=one;
	long long c=zero;
	for(;bchk|c;bchk>>=one){
		long long cx=((a&b&msk)^(b&c)^(c&a));
		a^=(b&msk)^c;
		c=cx;
		c<<=one;
		msk<<=one;
	}
	return a;
}
long long sub(long long a,long long b){
	return add(~add(~zero,b),a);
}

const long long mo=one|two|four|one<<nine|one<<add(one,ten)|one<<add(four,ten)|one<<add(five,ten)|one<<add(seven,ten)|one<<add(nine,ten)|one<<add(ten,ten)|one<<add(add(three,ten),ten)|one<<add(add(four,ten),ten)|one<<add(add(five,ten),ten)|one<<add(add(seven,ten),ten)|one<<add(add(eight,ten),ten)|one<<add(add(nine,ten),ten);

long long mul(long long a,long long b){
	long long ret=zero;
	for(;b;b>>=one){
		if(b&one)ret=add(ret,a);
		a=add(a,a);
	}
	return ret;
}

long long _div(long long a,long long b){
	long long ret=zero;
	for(;a>=b;){
		long long i=zero;
		for(;a>=b<<i;i=add(i,one));
		i=sub(i,one);
		a=sub(a,b<<i);
		ret|=one<<i;
	}
	return ret;
}

long long mod(long long a,long long b){
	return sub(a,mul(_div(a,b),b));
}

V M(const V &x,const V &y){
	long long a=zero,b=one,c=two,d=three,e=four;
	V z(e);
	z[a]=mod(add(mul(x[a],y[a]),mul(x[b],y[c])),mo);
	z[b]=mod(add(mul(x[a],y[b]),mul(x[b],y[d])),mo);
	z[c]=mod(add(mul(x[c],y[a]),mul(x[d],y[c])),mo);
	z[d]=mod(add(mul(x[c],y[b]),mul(x[d],y[d])),mo);
	return z;
}

int main(){
	long long T;
	for(cin>>T;T;T=sub(T,one)){
		long long N;
		cin>>N;N=sub(N,one);
		V e={one,zero,zero,one};
		V x={one,one,one,zero};
		for(;N;N>>=one){
			if(N&one)e=M(e,x);
			x=M(x,x);
		}
		cout<<mod(add(e[zero],e[one]<<one),mo)<<endl;
	}
}
